jwst/emicorr/emicorr.py

```diff
@@ -0,0 +1,792 @@
+
+    user_supplied_reffile : str
+        Reference file supplied by the user
+
+    nints_to_phase : int
+        Number of integrations to phase
+
+    nbins : int
+        Number of bins in one phased wave
+
+    scale_reference : bool
+        If True, the reference wavelength will be scaled to the data's phase amplitude
+
+    Returns
+    -------
+    output_model : JWST data model
+        input science data model which has been emi-corrected
+    """
+    # get the subarray case and other info
+    detector = input_model.meta.instrument.detector
+    subarray = input_model.meta.subarray.name
+    readpatt = input_model.meta.exposure.readpatt
+    xsize = input_model.meta.subarray.xsize   # SUBSIZE1 keyword
+    xstart = input_model.meta.subarray.xstart   # SUBSTRT1 keyword
+
+    # get the subarray case from either the ref file or set default values
+    freqs_numbers = []
+    if emicorr_model is not None:
+        log.info('Using reference file to get subarray case.')
+        subname, rowclocks, frameclocks, freqs2correct = get_subarcase(emicorr_model, subarray, readpatt, detector)
+        reference_wave_list = []
+        for fnme in freqs2correct:
+            freq, ref_wave = get_frequency_info(emicorr_model, fnme)
+            freqs_numbers.append(freq)
+            reference_wave_list.append(ref_wave)
+    else:
+        log.info('Using default subarray case corrections.')
+        subname, rowclocks, frameclocks, freqs2correct = get_subarcase(default_subarray_cases, subarray, readpatt, detector)
+        for fnme in freqs2correct:
+            freq = get_frequency_info(default_emi_freqs, fnme)
+            freqs_numbers.append(freq)
+
+    if rowclocks is None:
+        # no subarray match found, print to log and skip correction
+        return subname
+
+    # get the number of samples, 10us sample times per pixel (1 for fastmode, 9 for slowmode)
+    nsamples = input_model.meta.exposure.nsamples
+
+    # Initialize the output model as a copy of the input
+    output_model = input_model.copy()
+
+    # create the dictionary to store the frequencies and corresponding phase amplidues
+    if save_intermediate_results and save_onthefly_reffile is not None:
+        freq_pa_dict = {'frequencies': {}, 'subarray_cases': {}}
+
+    # Loop over the frequencies to correct
+    log.info('With configuration: Subarray={}, Read_pattern={}, Detector={}'.format(subarray, readpatt, detector))
+    log.info('Will correct data for the following {} frequencies: '.format(len(freqs2correct)))
+    log.info('   {}'.format(freqs2correct))
+    for fi, frequency_name in enumerate(freqs2correct):
+        frequency = freqs_numbers[fi]
+        log.info('Correcting for frequency: {} Hz  ({} out of {})'.format(frequency, fi+1, len(freqs2correct)))
+
+        # Read image data and set up some variables
+        orig_data = output_model.data
+        data = orig_data.copy()
```

Comment:
Because `output_model.data` is already a copy of the input, is it really necessary to make another copy of the `data` array?

Reply:
yep, this is necessary because the data array is modified along the step to calculate the noise, but then you want to subtract it from the original array